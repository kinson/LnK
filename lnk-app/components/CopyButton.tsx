import React, { ReactElement, useRef, useState } from 'react';

interface Props {
  disabled: boolean;
  link: string | null;
}

export default function ShortenButton({ link, disabled }: Props): ReactElement {
  const inputRef = useRef(null);
  const [copied, setCopied] = useState<boolean>(false);

  const buttonText = copied ? 'Copied!' : '& Copy';

  const baseStyle =
    'bg-violet ml-0.5 px-3 py-2 rounded-tr-md rounded-br-md text-white';

  const style = disabled
    ? `${baseStyle} bg-opacity-60 text-opacity-70`
    : baseStyle;

  const copyLinkToClipboard = () => {
    if (inputRef.current !== null && link !== null) {
      inputRef.current.focus();
      inputRef.current.select();

      if (document.execCommand('copy')) {
        setCopied(true);
      }
    }
  };

  return (
    <>
      <button
        disabled={disabled}
        onClick={copyLinkToClipboard}
        type="button"
        className={style}
      >
        {buttonText}
      </button>
      <input
        ref={inputRef}
        value={link ?? ''}
        className="h-0 w-0 focus:outline-none"
        readOnly
      />
    </>
  );
}
