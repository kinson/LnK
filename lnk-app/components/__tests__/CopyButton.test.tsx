import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import CopyButton from '../CopyButton';

describe('CopyButton component tests', () => {
  it('should render the CopyButton component when it is not disabled', () => {
    const component = renderer.create(
      <CopyButton disabled={false} onClick={jest.fn()} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render the CopyButton component when it is disabled', () => {
    const component = renderer.create(
      <CopyButton disabled onClick={jest.fn()} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should call the onClick function when it is pressed', () => {
    const onClickMock = jest.fn();
    const component = render(
      <CopyButton disabled={false} onClick={onClickMock} />
    );

    fireEvent.click(component.queryAllByText(/& Copy/i)[0]);

    expect(onClickMock).toHaveBeenCalled();
  });
});
