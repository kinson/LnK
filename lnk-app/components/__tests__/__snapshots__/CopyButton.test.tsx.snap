// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton component tests should render the CopyButton component when it is disabled 1`] = `
<button
  className="bg-violet ml-0.5 px-3 py-2 rounded-tr-md rounded-br-md text-white bg-opacity-60 text-opacity-70"
  disabled={true}
  onClick={[MockFunction]}
  type="button"
>
  & Copy
</button>
`;

exports[`CopyButton component tests should render the CopyButton component when it is not disabled 1`] = `
<button
  className="bg-violet ml-0.5 px-3 py-2 rounded-tr-md rounded-br-md text-white"
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  & Copy
</button>
`;

exports[`CopyButton component tests should render the CopyButton component with a message 1`] = `
<button
  className="bg-violet ml-0.5 px-3 py-2 rounded-tr-md rounded-br-md text-white"
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  & Copy
</button>
`;
